src/combobox/ComboBox.ts

```diff
@@ -0,0 +1,431 @@
+	}
+
+	@onPropertiesChanged
+	protected onPropertiesChanged(evt: PropertiesChangeEvent<this, ComboBoxProperties>) {
+		if (!this.registry || includes(evt.changedPropertyKeys, 'customResultItem') || includes(evt.changedPropertyKeys, 'customResultMenu')) {
+			this.registry = this._createRegistry();
+		}
+	}
+
+	protected scrollIntoView(element: HTMLElement) {
+		const menu = <HTMLElement> element.parentElement;
+		// Scroll menu up so top of highlighted result aligns with top of menu container
+		if (element.offsetTop - menu.scrollTop < 0) {
+			menu.scrollTop = element.offsetTop;
+		}
+		// Scroll menu down so bottom of highlighted result aligns with bottom of menu container
+		else if ((element.offsetTop - menu.scrollTop + element.offsetHeight) > menu.clientHeight) {
+			menu.scrollTop = element.offsetTop - menu.clientHeight + element.offsetHeight;
+		}
+	}
+
+	protected updateSelectedIndex(operation?: Operation, index?: number) {
+		const { results } = this.properties;
+
+		if (!results || results.length === 0 || !this._open) {
+			return;
+		}
+
+		if (index !== undefined) {
+			this._selectedIndex = index;
+		}
+		else if (operation !== undefined) {
+			if (operation === Operation.increase || operation === Operation.decrease) {
+				const total = results.length;
+				const base = operation === Operation.increase ? 0 : -1;
+				const current = this._selectedIndex !== undefined ? this._selectedIndex + operation : base;
+				this._selectedIndex = (current + total) % total;
+			}
+			else {
+				this._selectedIndex = undefined;
+			}
+			this._direction = operation;
+		}
+	}
+
+	render(): DNode {
+		const {
+			clearable,
+			disabled,
+			formId,
+			inputProperties = {},
+			invalid,
+			label,
+			readOnly,
+			required,
+			results = [],
+			value = ''
+		} = this.properties;
+
+		const menuId = uuid();
+		const menu = this._renderMenu(results, menuId);
+		this._notifyMenuChange();
+
+		this._wasOpen = this._open;
+
+		const input = w(TextInput, {
+			bind: this,
+			classes: this.classes(clearable ? css.clearable : null),
+			controls: menuId,
+			disabled,
+			invalid,
+			readOnly,
+			required,
+			value,
+			onBlur: this._onInputBlur,
+			onFocus: this._onInputFocus,
+			onInput: this._onInput,
+			onKeyDown: this._onInputKeyDown,
+			overrideClasses: css,
+			...inputProperties
+		});
+
+		return v('div', {
+			afterCreate: this._afterCreate,
+			afterUpdate: this._afterUpdate,
+			'aria-expanded': this._open ? 'true' : 'false',
+			'aria-haspopup': 'true',
+			'aria-readonly': readOnly ? 'true' : 'false',
+			'aria-required': required ? 'true' : 'false',
+			bind: this,
+			classes: this.classes(css.root),
+			role: 'root'
+		}, [
+			label ? w(Label, {
+				bind: this,
+				formId,
+				label
+			}, [ input ]) : input,
+			clearable ? v('button', {
+				'aria-controls': menuId,
+				bind: this,
+				classes: this.classes(css.clear),
+				innerHTML: 'clear combo box',
+				onclick: this._onClearClick
+			}) : null,
+			v('button', {
```

Comment:
Just realized -- button should probably also be disabled when the combobox is disabled. Right now you can still tab to it (and the clearable button too, I'm guessing).

Also, now that the label wraps the text input, I think you need to check positioning of the button. Since the label text can go either before or after the input I'm not sure accurate positioning is possible in the current config. I might take back my previous suggestion and have the label wrap everything again, but add a `for` attribute pointing at the text input id.